pkg/networkmanager/firewall.jsx

```diff
@@ -171,6 +193,136 @@ class AddServicesBody extends React.Component {
+            if (m.length > 5 && m[5])
+                new_port['description'] = m[5].trim();
+            if (ret[m[1]])
+                ret[m[1]]['type'].push(new_port['type'][0]);
+            else
+                ret[m[1]] = new_port;
+            if (ret[m[2]])
+                ret[m[2]]['type'].push(new_port['type'][0]);
+            else
+                ret[m[2]] = new_port;
+        }
+        return ret;
+    }
+
+    setName(event) {
+        this.setState({
+            custom_name: event.target.value,
+            generate_name: false,
+        });
+    }
+
+    getName(port) {
+        let known = this.state.avail_services[port];
+        if (known)
+            return known['name'];
+        else
+            return port;
+    }
+
+    validate(event) {
+        let ports = event.target.value.split(',');
+        let error = "";
+        let targets = ['tcp', 'custom_tcp_ports', 'tcp_error', 'custom_tcp_value'];
+        if (event.target.id === "udp-ports")
+            targets = ['udp', 'custom_udp_ports', 'udp_error', 'custom_udp_value'];
+
+        let new_ports = [];
+        let self = this;
+        ports.forEach(function(port) {
+            port = port.trim();
+            if (!port)
+                return;
+            let p = self.state.avail_services[port.trim()];
+            if (!p) {
+                let num_p = Number(port);
+                if (isNaN(num_p))
+                    error = "Unknown service name";
+                else if (num_p <= 0 || num_p > 65535)
+                    error = "Not valid port number";
+                else
+                    new_ports.push(port);
+            } else if (p['type'].indexOf(targets[0]) < 0)
+                error = "Not matching port number and type";
+            else {
+                new_ports.push(p.port);
+            }
+        });
+        let newState = { [targets[1]]: new_ports };
+        newState[targets[2]] = error;
+        newState[targets[3]] = event.target.value;
+
+        let name = this.state.custom_name;
+        let all_ports = new_ports.concat(this.state.custom_udp_ports);
+        if (event.target.id === "udp-ports")
+            all_ports = this.state.custom_tcp_ports.concat(new_ports);
+
+        if (this.state.generate_name) {
+            name = "";
+            if (all_ports.length === 1) {
+                let known = this.state.avail_services[all_ports[0]];
+                if (known)
+                    name = known['description'] || known['name'];
+                else
+                    name = all_ports[0];
+            } else if (all_ports.length > 1) {
+                let max = all_ports.length === 2 ? 2 : 3;
+                for (var i = 0; i < max; i++)
+                    name += this.getName(all_ports[i]) + ", ";
+                if (all_ports.length > 3)
+                    name += "etc.";
+                else
+                    name = name.slice(0, -2);
+            }
+        }
```

Comment:
Please move this out into a helper function. Also, it seems easier to assemble this as a list and use `.join(', ')` in the end. Finally, I'm fairly sure that "etc."  needs to be i18n'ed. How about "…" instead? We use that in several places.

Reply:
I am fine with `...` if it can be in the service name (I think it can).